File: starters/L1.2/try_pattern.py
"""
L1.2 starter — run the same task through five different prompt patterns.

Edit INPUT below to be a real piece of messy text from your work
(Slack thread, meeting notes, customer email — anonymise if needed).

Each PATTERN is a different prompt design. Run them all, compare the
outputs. Save your winner to ../../cohort-2026Q3/<your-name>/L1.2-prompt-v1.txt

Run:
  python try_pattern.py 1     # runs pattern 1 only
  python try_pattern.py all   # runs all five
"""
import sys
import anthropic

client = anthropic.Anthropic()
MODEL = "claude-haiku-4-5"

INPUT = """
PASTE YOUR MESSY TEXT HERE.

A long Slack thread, a meeting note, a customer email, a JIRA comment.
Anonymise it if needed.
"""

EXAMPLE_INPUT = """We discussed shipping the auth flow this Friday. Ana
will own the backend changes and asks for an extra reviewer. We still
don't know if SSO is in scope — leaving that to next standup."""

EXAMPLE_OUTPUT = {
    "decisions": ["Ship auth flow this Friday"],
    "owners": ["Ana — backend changes"],
    "open_questions": ["Is SSO in scope?"],
}

PATTERNS = {
    "1": {
        "label": "Bare instruction",
        "system": None,
        "user": f"Summarise the decisions, owners, and open questions from this text: {INPUT}",
        "prefill": None,
    },
    "2": {
        "label": "Instruction + format spec",
        "system": None,
        "user": (
            "Extract decisions, owners, and open questions from the text below. "
            "Return strict JSON with keys decisions (list of strings), owners "
            "(list of strings), open_questions (list of strings). No prose.\n\n"
            f"TEXT:\n{INPUT}"
        ),
        "prefill": None,
    },
    "3": {
        "label": "Few-shot example",
        "system": None,
        "user": (
            "Extract decisions, owners, and open questions as strict JSON.\n\n"
            f"EXAMPLE INPUT:\n{EXAMPLE_INPUT}\n\n"
            f"EXAMPLE OUTPUT:\n{EXAMPLE_OUTPUT}\n\n"
            f"NOW DO THE SAME FOR THIS INPUT:\n{INPUT}"
        ),
        "prefill": None,
    },
    "4": {
        "label": "System prompt + role",
        "system": (
            "You are an analyst who extracts structured information from "
            "meeting notes. You always return strict JSON with keys "
            "decisions, owners, open_questions. You never add prose."
        ),
        "user": INPUT,
        "prefill": None,
    },
    "5": {
        "label": "System prompt + prefill",
        "system": (
            "You are an analyst who extracts structured information from "
            "meeting notes. Return strict JSON with keys decisions, owners, "
            "open_questions."
        ),
        "user": INPUT,
        "prefill": '{"decisions": [',
    },
}


def run(pattern_id: str) -> None:
    p = PATTERNS[pattern_id]
    print(f"\n=== Pattern {pattern_id}: {p['label']} ===")
    messages = [{"role": "user", "content": p["user"]}]
    if p["prefill"]:
        messages.append({"role": "assistant", "content": p["prefill"]})
    kwargs = {
        "model": MODEL,
        "max_tokens": 600,
        "messages": messages,
    }
    if p["system"]:
        kwargs["system"] = p["system"]
    resp = client.messages.create(**kwargs)
    out = (p["prefill"] or "") + resp.content[0].text
    print(out)
    print(f"--- in: {resp.usage.input_tokens}, out: {resp.usage.output_tokens}")


def main() -> None:
    if len(sys.argv) < 2:
        print("usage: python try_pattern.py [1|2|3|4|5|all]")
        sys.exit(1)
    if sys.argv[1] == "all":
        for k in PATTERNS:
            run(k)
    else:
        run(sys.argv[1])


if __name__ == "__main__":
    main()

File: starters/L1.2/try_pattern.ts
/**
 * L1.2 starter -- run the same task through five different prompt patterns.
 *
 * Edit INPUT below to be a real piece of messy text from your work
 * (Teams chat, meeting notes, customer email -- anonymise if needed).
 *
 * Each PATTERN is a different prompt design. Run them all, compare the
 * outputs. Save your winner to ../../cohort-2026Q3/<your-name>/L1.2-prompt-v1.txt
 *
 * Run:
 *   npx ts-node try_pattern.ts 1     # runs pattern 1 only
 *   npx ts-node try_pattern.ts all   # runs all five
 */
import Anthropic from "@anthropic-ai/sdk";

const client = new Anthropic();
const MODEL = "claude-haiku-4-5";

const INPUT = `
PASTE YOUR MESSY TEXT HERE.

A long Teams chat, a meeting note, a customer email, a JIRA comment.
Anonymise it if needed.
`;

const EXAMPLE_INPUT = `We discussed shipping the auth flow this Friday. Ana
will own the backend changes and asks for an extra reviewer. We still
don't know if SSO is in scope -- leaving that to next standup.`;

const EXAMPLE_OUTPUT = {
  decisions: ["Ship auth flow this Friday"],
  owners: ["Ana -- backend changes"],
  open_questions: ["Is SSO in scope?"],
};

const PATTERNS: Record<string, { label: string; system: string | null; user: string; prefill: string | null }> = {
  "1": {
    label: "Bare instruction",
    system: null,
    user: `Summarise the decisions, owners, and open questions from this text: ${INPUT}`,
    prefill: null,
  },
  "2": {
    label: "Instruction + format spec",
    system: null,
    user:
      "Extract decisions, owners, and open questions from the text below. " +
      "Return strict JSON with keys decisions (list of strings), owners " +
      "(list of strings), open_questions (list of strings). No prose.\n\n" +
      `TEXT:\n${INPUT}`,
    prefill: null,
  },
  "3": {
    label: "Few-shot example",
    system: null,
    user:
      "Extract decisions, owners, and open questions as strict JSON.\n\n" +
      `EXAMPLE INPUT:\n${EXAMPLE_INPUT}\n\n` +
      `EXAMPLE OUTPUT:\n${JSON.stringify(EXAMPLE_OUTPUT)}\n\n` +
      `NOW DO THE SAME FOR THIS INPUT:\n${INPUT}`,
    prefill: null,
  },
  "4": {
    label: "System prompt + role",
    system:
      "You are an analyst who extracts structured information from " +
      "meeting notes. You always return strict JSON with keys " +
      "decisions, owners, open_questions. You never add prose.",
    user: INPUT,
    prefill: null,
  },
  "5": {
    label: "System prompt + prefill",
    system:
      "You are an analyst who extracts structured information from " +
      "meeting notes. Return strict JSON with keys decisions, owners, " +
      "open_questions.",
    user: INPUT,
    prefill: '{"decisions": [',
  },
};

async function run(patternId: string): Promise<void> {
  const p = PATTERNS[patternId];
  console.log(`\n=== Pattern ${patternId}: ${p.label} ===`);

  const messages: Anthropic.Messages.MessageParam[] = [
    { role: "user", content: p.user },
  ];
  if (p.prefill) {
    messages.push({ role: "assistant", content: p.prefill });
  }

  const params: Anthropic.Messages.MessageCreateParamsNonStreaming = {
    model: MODEL,
    max_tokens: 600,
    messages,
  };
  if (p.system) params.system = p.system;

  const resp = await client.messages.create(params);
  const raw = resp.content[0].type === "text" ? resp.content[0].text : "";
  const out = (p.prefill ?? "") + raw;
  console.log(out);
  console.log(`--- in: ${resp.usage.input_tokens}, out: ${resp.usage.output_tokens}`);
}

(async () => {
  const arg = process.argv[2];
  if (!arg) {
    console.log("usage: npx ts-node try_pattern.ts [1|2|3|4|5|all]");
    process.exit(1);
  }
  if (arg === "all") {
    for (const k of Object.keys(PATTERNS)) await run(k);
  } else {
    await run(arg);
  }
})();
